{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# An IPerl helper for `perldoc` HTML\n",
    "\n",
    "author: Zaki Mughal\n",
    "\n",
    "date: 2015-02-11\n",
    "\n",
    "This notebook demonstrates how to add a helper function that reads in the output of `perldoc` and then displays it as HTML within the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<html><head><title>Try::Tiny</title>\n",
       "<meta http-equiv=\"Content-Type\" content=\"text/html; charset=ISO-8859-1\" >\n",
       "</head>\n",
       "<body class='pod'>\n",
       "<!--\n",
       "  generated by Pod::Simple::HTML v3.28,\n",
       "  using Pod::Simple::PullParser v3.28,\n",
       "  under Perl v5.016003 at Wed Feb 11 17:36:38 2015 GMT.\n",
       "\n",
       " If you want to change this HTML document, you probably shouldn't do that\n",
       "   by changing it directly.  Instead, see about changing the calling options\n",
       "   to Pod::Simple::HTML, and/or subclassing Pod::Simple::HTML,\n",
       "   then reconverting this document from the Pod source.\n",
       "   When in doubt, email the author of Pod::Simple::HTML for advice.\n",
       "   See 'perldoc Pod::Simple::HTML' for more info.\n",
       "\n",
       "-->\n",
       "\n",
       "<!-- start doc -->\n",
       "<a name='___top' class='dummyTopAnchor' ></a>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"NAME\"\n",
       ">NAME</a></h1>\n",
       "\n",
       "<p>Try::Tiny - minimal try/catch with proper preservation of $@</p>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"VERSION\"\n",
       ">VERSION</a></h1>\n",
       "\n",
       "<p>version 0.22</p>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"SYNOPSIS\"\n",
       ">SYNOPSIS</a></h1>\n",
       "\n",
       "<p>You can use Try::Tiny&#39;s <code>try</code> and <code>catch</code> to expect and handle exceptional conditions,\n",
       "avoiding quirks in Perl and common mistakes:</p>\n",
       "\n",
       "<pre>  # handle errors with a catch handler\n",
       "  try {\n",
       "    die &#34;foo&#34;;\n",
       "  } catch {\n",
       "    warn &#34;caught error: $_&#34;; # not $@\n",
       "  };</pre>\n",
       "\n",
       "<p>You can also use it like a standalone <code>eval</code> to catch and ignore any error conditions. Obviously, this is an extreme measure not to be undertaken lightly:</p>\n",
       "\n",
       "<pre>  # just silence errors\n",
       "  try {\n",
       "    die &#34;foo&#34;;\n",
       "  };</pre>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"DESCRIPTION\"\n",
       ">DESCRIPTION</a></h1>\n",
       "\n",
       "<p>This module provides bare bones <code>try</code>/<code>catch</code>/<code>finally</code> statements that are designed to minimize common mistakes with eval blocks, and NOTHING else.</p>\n",
       "\n",
       "<p>This is unlike <a href=\"http://search.cpan.org/perldoc?TryCatch\" class=\"podlinkpod\"\n",
       ">TryCatch</a> which provides a nice syntax and avoids adding another call stack layer, and supports calling <code>return</code> from the <code>try</code> block to return from the parent subroutine. These extra features come at a cost of a few dependencies, namely <a href=\"http://search.cpan.org/perldoc?Devel%3A%3ADeclare\" class=\"podlinkpod\"\n",
       ">Devel::Declare</a> and <a href=\"http://search.cpan.org/perldoc?Scope%3A%3AUpper\" class=\"podlinkpod\"\n",
       ">Scope::Upper</a> which are occasionally problematic, and the additional catch filtering uses <a href=\"http://search.cpan.org/perldoc?Moose\" class=\"podlinkpod\"\n",
       ">Moose</a> type constraints which may not be desirable either.</p>\n",
       "\n",
       "<p>The main focus of this module is to provide simple and reliable error handling for those having a hard time installing <a href=\"http://search.cpan.org/perldoc?TryCatch\" class=\"podlinkpod\"\n",
       ">TryCatch</a>, but who still want to write correct <code>eval</code> blocks without 5 lines of boilerplate each time.</p>\n",
       "\n",
       "<p>It&#39;s designed to work as correctly as possible in light of the various pathological edge cases (see <a href=\"#BACKGROUND\" class=\"podlinkpod\"\n",
       ">&#34;BACKGROUND&#34;</a>) and to be compatible with any style of error values (simple strings, references, objects, overloaded objects, etc).</p>\n",
       "\n",
       "<p>If the <code>try</code> block dies, it returns the value of the last statement executed in the <code>catch</code> block, if there is one. Otherwise, it returns <code>undef</code> in scalar context or the empty list in list context. The following examples all assign <code>&#34;bar&#34;</code> to <code>$x</code>:</p>\n",
       "\n",
       "<pre>  my $x = try { die &#34;foo&#34; } catch { &#34;bar&#34; };\n",
       "  my $x = try { die &#34;foo&#34; } || { &#34;bar&#34; };\n",
       "  my $x = (try { die &#34;foo&#34; }) // { &#34;bar&#34; };\n",
       "\n",
       "  my $x = eval { die &#34;foo&#34; } || &#34;bar&#34;;</pre>\n",
       "\n",
       "<p>You can add <code>finally</code> blocks, yielding the following:</p>\n",
       "\n",
       "<pre>  my $x;\n",
       "  try { die &#39;foo&#39; } finally { $x = &#39;bar&#39; };\n",
       "  try { die &#39;foo&#39; } catch { warn &#34;Got a die: $_&#34; } finally { $x = &#39;bar&#39; };</pre>\n",
       "\n",
       "<p><code>finally</code> blocks are always executed making them suitable for cleanup code which cannot be handled using local. You can add as many <code>finally</code> blocks to a given <code>try</code> block as you like.</p>\n",
       "\n",
       "<p>Note that adding a <code>finally</code> block without a preceding <code>catch</code> block suppresses any errors. This behaviour is consistent with using a standalone <code>eval</code>, but it is not consistent with <code>try</code>/<code>finally</code> patterns found in other programming languages, such as Java, Python, Javascript or C#. If you learnt the <code>try</code>/<code>finally</code> pattern from one of these languages, watch out for this.</p>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"EXPORTS\"\n",
       ">EXPORTS</a></h1>\n",
       "\n",
       "<p>All functions are exported by default using <a href=\"http://search.cpan.org/perldoc?Exporter\" class=\"podlinkpod\"\n",
       ">Exporter</a>.</p>\n",
       "\n",
       "<p>If you need to rename the <code>try</code>, <code>catch</code> or <code>finally</code> keyword consider using <a href=\"http://search.cpan.org/perldoc?Sub%3A%3AImport\" class=\"podlinkpod\"\n",
       ">Sub::Import</a> to get <a href=\"http://search.cpan.org/perldoc?Sub%3A%3AExporter\" class=\"podlinkpod\"\n",
       ">Sub::Exporter</a>&#39;s flexibility.</p>\n",
       "\n",
       "<dl>\n",
       "<dt><a name=\"try_(&#38;;@)\"\n",
       ">try (&#38;;@)</a></dt>\n",
       "\n",
       "<dd>\n",
       "<p>Takes one mandatory <code>try</code> subroutine, an optional <code>catch</code> subroutine and <code>finally</code> subroutine.</p>\n",
       "\n",
       "<p>The mandatory subroutine is evaluated in the context of an <code>eval</code> block.</p>\n",
       "\n",
       "<p>If no error occurred the value from the first block is returned, preserving list/scalar context.</p>\n",
       "\n",
       "<p>If there was an error and the second subroutine was given it will be invoked with the error in <code>$_</code> (localized) and as that block&#39;s first and only argument.</p>\n",
       "\n",
       "<p><code>$@</code> does <b>not</b> contain the error. Inside the <code>catch</code> block it has the same value it had before the <code>try</code> block was executed.</p>\n",
       "\n",
       "<p>Note that the error may be false, but if that happens the <code>catch</code> block will still be invoked.</p>\n",
       "\n",
       "<p>Once all execution is finished then the <code>finally</code> block, if given, will execute.</p>\n",
       "\n",
       "<dt><a name=\"catch_(&#38;;@)\"\n",
       ">catch (&#38;;@)</a></dt>\n",
       "\n",
       "<dd>\n",
       "<p>Intended to be used in the second argument position of <code>try</code>.</p>\n",
       "\n",
       "<p>Returns a reference to the subroutine it was given but blessed as <code>Try::Tiny::Catch</code> which allows try to decode correctly what to do with this code reference.</p>\n",
       "\n",
       "<pre>  catch { ... }</pre>\n",
       "\n",
       "<p>Inside the <code>catch</code> block the caught error is stored in <code>$_</code>, while previous value of <code>$@</code> is still available for use. This value may or may not be meaningful depending on what happened before the <code>try</code>, but it might be a good idea to preserve it in an error stack.</p>\n",
       "\n",
       "<p>For code that captures <code>$@</code> when throwing new errors (i.e. <a href=\"http://search.cpan.org/perldoc?Class%3A%3AThrowable\" class=\"podlinkpod\"\n",
       ">Class::Throwable</a>), you&#39;ll need to do:</p>\n",
       "\n",
       "<pre>  local $@ = $_;</pre>\n",
       "\n",
       "<dt><a name=\"finally_(&#38;;@)\"\n",
       ">finally (&#38;;@)</a></dt>\n",
       "\n",
       "<dd>\n",
       "<pre>  try     { ... }\n",
       "  catch   { ... }\n",
       "  finally { ... };</pre>\n",
       "\n",
       "<p>Or</p>\n",
       "\n",
       "<pre>  try     { ... }\n",
       "  finally { ... };</pre>\n",
       "\n",
       "<p>Or even</p>\n",
       "\n",
       "<pre>  try     { ... }\n",
       "  finally { ... }\n",
       "  catch   { ... };</pre>\n",
       "\n",
       "<p>Intended to be the second or third element of <code>try</code>. <code>finally</code> blocks are always executed in the event of a successful <code>try</code> or if <code>catch</code> is run. This allows you to locate cleanup code which cannot be done via <code>local()</code> e.g. closing a file handle.</p>\n",
       "\n",
       "<p>When invoked, the <code>finally</code> block is passed the error that was caught. If no error was caught, it is passed nothing. (Note that the <code>finally</code> block does not localize <code>$_</code> with the error, since unlike in a <code>catch</code> block, there is no way to know if <code>$_ == undef</code> implies that there were no errors.) In other words, the following code does just what you would expect:</p>\n",
       "\n",
       "<pre>  try {\n",
       "    die_sometimes();\n",
       "  } catch {\n",
       "    # ...code run in case of error\n",
       "  } finally {\n",
       "    if (@_) {\n",
       "      print &#34;The try block died with: @_\\n&#34;;\n",
       "    } else {\n",
       "      print &#34;The try block ran without error.\\n&#34;;\n",
       "    }\n",
       "  };</pre>\n",
       "\n",
       "<p><b>You must always do your own error handling in the <code>finally</code> block</b>. <code>Try::Tiny</code> will not do anything about handling possible errors coming from code located in these blocks.</p>\n",
       "\n",
       "<p>Furthermore <b>exceptions in <code>finally</code> blocks are not trappable and are unable to influence the execution of your program</b>. This is due to limitation of <code>DESTROY</code>-based scope guards, which <code>finally</code> is implemented on top of. This may change in a future version of Try::Tiny.</p>\n",
       "\n",
       "<p>In the same way <code>catch()</code> blesses the code reference this subroutine does the same except it bless them as <code>Try::Tiny::Finally</code>.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"BACKGROUND\"\n",
       ">BACKGROUND</a></h1>\n",
       "\n",
       "<p>There are a number of issues with <code>eval</code>.</p>\n",
       "\n",
       "<h2><a class='u'\n",
       "name=\"Clobbering_$@\"\n",
       ">Clobbering $@</a></h2>\n",
       "\n",
       "<p>When you run an <code>eval</code> block and it succeeds, <code>$@</code> will be cleared, potentially clobbering an error that is currently being caught.</p>\n",
       "\n",
       "<p>This causes action at a distance, clearing previous errors your caller may have not yet handled.</p>\n",
       "\n",
       "<p><code>$@</code> must be properly localized before invoking <code>eval</code> in order to avoid this issue.</p>\n",
       "\n",
       "<p>More specifically, <code>$@</code> is clobbered at the beginning of the <code>eval</code>, which also makes it impossible to capture the previous error before you die (for instance when making exception objects with error stacks).</p>\n",
       "\n",
       "<p>For this reason <code>try</code> will actually set <code>$@</code> to its previous value (the one available before entering the <code>try</code> block) in the beginning of the <code>eval</code> block.</p>\n",
       "\n",
       "<h2><a class='u'\n",
       "name=\"Localizing_$@_silently_masks_errors\"\n",
       ">Localizing $@ silently masks errors</a></h2>\n",
       "\n",
       "<p>Inside an <code>eval</code> block, <code>die</code> behaves sort of like:</p>\n",
       "\n",
       "<pre>  sub die {\n",
       "    $@ = $_[0];\n",
       "    return_undef_from_eval();\n",
       "  }</pre>\n",
       "\n",
       "<p>This means that if you were polite and localized <code>$@</code> you can&#39;t die in that scope, or your error will be discarded (printing &#34;Something&#39;s wrong&#34; instead).</p>\n",
       "\n",
       "<p>The workaround is very ugly:</p>\n",
       "\n",
       "<pre>  my $error = do {\n",
       "    local $@;\n",
       "    eval { ... };\n",
       "    $@;\n",
       "  };\n",
       "\n",
       "  ...\n",
       "  die $error;</pre>\n",
       "\n",
       "<h2><a class='u'\n",
       "name=\"$@_might_not_be_a_true_value\"\n",
       ">$@ might not be a true value</a></h2>\n",
       "\n",
       "<p>This code is wrong:</p>\n",
       "\n",
       "<pre>  if ( $@ ) {\n",
       "    ...\n",
       "  }</pre>\n",
       "\n",
       "<p>because due to the previous caveats it may have been unset.</p>\n",
       "\n",
       "<p><code>$@</code> could also be an overloaded error object that evaluates to false, but that&#39;s asking for trouble anyway.</p>\n",
       "\n",
       "<p>The classic failure mode is:</p>\n",
       "\n",
       "<pre>  sub Object::DESTROY {\n",
       "    eval { ... }\n",
       "  }\n",
       "\n",
       "  eval {\n",
       "    my $obj = Object-&#62;new;\n",
       "\n",
       "    die &#34;foo&#34;;\n",
       "  };\n",
       "\n",
       "  if ( $@ ) {\n",
       "\n",
       "  }</pre>\n",
       "\n",
       "<p>In this case since <code>Object::DESTROY</code> is not localizing <code>$@</code> but still uses <code>eval</code>, it will set <code>$@</code> to <code>&#34;&#34;</code>.</p>\n",
       "\n",
       "<p>The destructor is called when the stack is unwound, after <code>die</code> sets <code>$@</code> to <code>&#34;foo at Foo.pm line 42\\n&#34;</code>, so by the time <code>if ( $@ )</code> is evaluated it has been cleared by <code>eval</code> in the destructor.</p>\n",
       "\n",
       "<p>The workaround for this is even uglier than the previous ones. Even though we can&#39;t save the value of <code>$@</code> from code that doesn&#39;t localize, we can at least be sure the <code>eval</code> was aborted due to an error:</p>\n",
       "\n",
       "<pre>  my $failed = not eval {\n",
       "    ...\n",
       "\n",
       "    return 1;\n",
       "  };</pre>\n",
       "\n",
       "<p>This is because an <code>eval</code> that caught a <code>die</code> will always return a false value.</p>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"SHINY_SYNTAX\"\n",
       ">SHINY SYNTAX</a></h1>\n",
       "\n",
       "<p>Using Perl 5.10 you can use <a href=\"http://search.cpan.org/perldoc?perlsyn#Switch_statements\" class=\"podlinkpod\"\n",
       ">&#34;Switch statements&#34; in perlsyn</a>.</p>\n",
       "\n",
       "<p>The <code>catch</code> block is invoked in a topicalizer context (like a <code>given</code> block), but note that you can&#39;t return a useful value from <code>catch</code> using the <code>when</code> blocks without an explicit <code>return</code>.</p>\n",
       "\n",
       "<p>This is somewhat similar to Perl 6&#39;s <code>CATCH</code> blocks. You can use it to concisely match errors:</p>\n",
       "\n",
       "<pre>  try {\n",
       "    require Foo;\n",
       "  } catch {\n",
       "    when (/^Can&#39;t locate .*?\\.pm in \\@INC/) { } # ignore\n",
       "    default { die $_ }\n",
       "  };</pre>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"CAVEATS\"\n",
       ">CAVEATS</a></h1>\n",
       "\n",
       "<ul>\n",
       "<li><code>@_</code> is not available within the <code>try</code> block, so you need to copy your arglist. In case you want to work with argument values directly via <code>@_</code> aliasing (i.e. allow <code>$_[1] = &#34;foo&#34;</code>), you need to pass <code>@_</code> by reference:\n",
       "<pre>  sub foo {\n",
       "    my ( $self, @args ) = @_;\n",
       "    try { $self-&#62;bar(@args) }\n",
       "  }</pre>\n",
       "\n",
       "<p>or</p>\n",
       "\n",
       "<pre>  sub bar_in_place {\n",
       "    my $self = shift;\n",
       "    my $args = \\@_;\n",
       "    try { $_ = $self-&#62;bar($_) for @$args }\n",
       "  }</pre>\n",
       "</li>\n",
       "\n",
       "<li><code>return</code> returns from the <code>try</code> block, not from the parent sub (note that this is also how <code>eval</code> works, but not how <a href=\"http://search.cpan.org/perldoc?TryCatch\" class=\"podlinkpod\"\n",
       ">TryCatch</a> works):\n",
       "<pre>  sub parent_sub {\n",
       "    try {\n",
       "      die;\n",
       "    }\n",
       "    catch {\n",
       "      return;\n",
       "    };\n",
       "\n",
       "    say &#34;this text WILL be displayed, even though an exception is thrown&#34;;\n",
       "  }</pre>\n",
       "\n",
       "<p>Instead, you should capture the return value:</p>\n",
       "\n",
       "<pre>  sub parent_sub {\n",
       "    my $success = try {\n",
       "      die;\n",
       "      1;\n",
       "    };\n",
       "    return unless $success;\n",
       "\n",
       "    say &#34;This text WILL NEVER appear!&#34;;\n",
       "  }\n",
       "  # OR\n",
       "  sub parent_sub_with_catch {\n",
       "    my $success = try {\n",
       "      die;\n",
       "      1;\n",
       "    }\n",
       "    catch {\n",
       "      # do something with $_\n",
       "      return undef; #see note\n",
       "    };\n",
       "    return unless $success;\n",
       "\n",
       "    say &#34;This text WILL NEVER appear!&#34;;\n",
       "  }</pre>\n",
       "\n",
       "<p>Note that if you have a <code>catch</code> block, it must return <code>undef</code> for this to work, since if a <code>catch</code> block exists, its return value is returned in place of <code>undef</code> when an exception is thrown.</p>\n",
       "</li>\n",
       "\n",
       "<li><code>try</code> introduces another caller stack frame. <a href=\"http://search.cpan.org/perldoc?Sub%3A%3AUplevel\" class=\"podlinkpod\"\n",
       ">Sub::Uplevel</a> is not used. <a href=\"http://search.cpan.org/perldoc?Carp\" class=\"podlinkpod\"\n",
       ">Carp</a> will not report this when using full stack traces, though, because <code>%Carp::Internal</code> is used. This lack of magic is considered a feature.</li>\n",
       "\n",
       "<li>The value of <code>$_</code> in the <code>catch</code> block is not guaranteed to be the value of the exception thrown (<code>$@</code>) in the <code>try</code> block. There is no safe way to ensure this, since <code>eval</code> may be used unhygenically in destructors. The only guarantee is that the <code>catch</code> will be called if an exception is thrown.</li>\n",
       "\n",
       "<li>The return value of the <code>catch</code> block is not ignored, so if testing the result of the expression for truth on success, be sure to return a false value from the <code>catch</code> block:\n",
       "<pre>  my $obj = try {\n",
       "    MightFail-&#62;new;\n",
       "  } catch {\n",
       "    ...\n",
       "\n",
       "    return; # avoid returning a true value;\n",
       "  };\n",
       "\n",
       "  return unless $obj;</pre>\n",
       "</li>\n",
       "\n",
       "<li><code>$SIG{__DIE__}</code> is still in effect.\n",
       "<p>Though it can be argued that <code>$SIG{__DIE__}</code> should be disabled inside of <code>eval</code> blocks, since it isn&#39;t people have grown to rely on it. Therefore in the interests of compatibility, <code>try</code> does not disable <code>$SIG{__DIE__}</code> for the scope of the error throwing code.</p>\n",
       "</li>\n",
       "\n",
       "<li>Lexical <code>$_</code> may override the one set by <code>catch</code>.\n",
       "<p>For example Perl 5.10&#39;s <code>given</code> form uses a lexical <code>$_</code>, creating some confusing behavior:</p>\n",
       "\n",
       "<pre>  given ($foo) {\n",
       "    when (...) {\n",
       "      try {\n",
       "        ...\n",
       "      } catch {\n",
       "        warn $_; # will print $foo, not the error\n",
       "        warn $_[0]; # instead, get the error like this\n",
       "      }\n",
       "    }\n",
       "  }</pre>\n",
       "\n",
       "<p>Note that this behavior was changed once again in <a href=\"https://metacpan.org/module/perldelta#given-now-aliases-the-global-_\" class=\"podlinkurl\"\n",
       ">Perl5 version 18 </a>. However, since the entirety of lexical <code>$_</code> is now <a href=\"https://metacpan.org/module/perldelta#Lexical-_-is-now-experimental\" class=\"podlinkurl\"\n",
       ">considired experimental </a>, it is unclear whether the new version 18 behavior is final.</p>\n",
       "</li>\n",
       "</ul>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"SEE_ALSO\"\n",
       ">SEE ALSO</a></h1>\n",
       "\n",
       "<dl>\n",
       "<dt><a name=\"TryCatch\"\n",
       "><a href=\"http://search.cpan.org/perldoc?TryCatch\" class=\"podlinkpod\"\n",
       ">TryCatch</a></a></dt>\n",
       "\n",
       "<dd>\n",
       "<p>Much more feature complete, more convenient semantics, but at the cost of implementation complexity.</p>\n",
       "\n",
       "<dt><a name=\"autodie\"\n",
       "><a href=\"http://search.cpan.org/perldoc?autodie\" class=\"podlinkpod\"\n",
       ">autodie</a></a></dt>\n",
       "\n",
       "<dd>\n",
       "<p>Automatic error throwing for builtin functions and more. Also designed to work well with <code>given</code>/<code>when</code>.</p>\n",
       "\n",
       "<dt><a name=\"Throwable\"\n",
       "><a href=\"http://search.cpan.org/perldoc?Throwable\" class=\"podlinkpod\"\n",
       ">Throwable</a></a></dt>\n",
       "\n",
       "<dd>\n",
       "<p>A lightweight role for rolling your own exception classes.</p>\n",
       "\n",
       "<dt><a name=\"Error\"\n",
       "><a href=\"http://search.cpan.org/perldoc?Error\" class=\"podlinkpod\"\n",
       ">Error</a></a></dt>\n",
       "\n",
       "<dd>\n",
       "<p>Exception object implementation with a <code>try</code> statement. Does not localize <code>$@</code>.</p>\n",
       "\n",
       "<dt><a name=\"Exception::Class::TryCatch\"\n",
       "><a href=\"http://search.cpan.org/perldoc?Exception%3A%3AClass%3A%3ATryCatch\" class=\"podlinkpod\"\n",
       ">Exception::Class::TryCatch</a></a></dt>\n",
       "\n",
       "<dd>\n",
       "<p>Provides a <code>catch</code> statement, but properly calling <code>eval</code> is your responsibility.</p>\n",
       "\n",
       "<p>The <code>try</code> keyword pushes <code>$@</code> onto an error stack, avoiding some of the issues with <code>$@</code>, but you still need to localize to prevent clobbering.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"LIGHTNING_TALK\"\n",
       ">LIGHTNING TALK</a></h1>\n",
       "\n",
       "<p>I gave a lightning talk about this module, you can see the slides (Firefox only):</p>\n",
       "\n",
       "<p><a href=\"http://web.archive.org/web/20100628040134/http://nothingmuch.woobling.org/talks/takahashi.xul\" class=\"podlinkurl\"\n",
       ">http://web.archive.org/web/20100628040134/http://nothingmuch.woobling.org/talks/takahashi.xul</a></p>\n",
       "\n",
       "<p>Or read the source:</p>\n",
       "\n",
       "<p><a href=\"http://web.archive.org/web/20100305133605/http://nothingmuch.woobling.org/talks/yapc_asia_2009/try_tiny.yml\" class=\"podlinkurl\"\n",
       ">http://web.archive.org/web/20100305133605/http://nothingmuch.woobling.org/talks/yapc_asia_2009/try_tiny.yml</a></p>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"VERSION_CONTROL\"\n",
       ">VERSION CONTROL</a></h1>\n",
       "\n",
       "<p><a href=\"http://github.com/doy/try-tiny/\" class=\"podlinkurl\"\n",
       ">http://github.com/doy/try-tiny/</a></p>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"AUTHORS\"\n",
       ">AUTHORS</a></h1>\n",
       "\n",
       "<ul>\n",
       "<li>Yuval Kogman &#60;nothingmuch@woobling.org&#62;</li>\n",
       "\n",
       "<li>Jesse Luehrs &#60;doy@tozt.net&#62;</li>\n",
       "</ul>\n",
       "\n",
       "<h1><a class='u'\n",
       "name=\"COPYRIGHT_AND_LICENSE\"\n",
       ">COPYRIGHT AND LICENSE</a></h1>\n",
       "\n",
       "<p>This software is Copyright (c) 2014 by Yuval Kogman.</p>\n",
       "\n",
       "<p>This is free software, licensed under:</p>\n",
       "\n",
       "<pre>  The MIT (X11) License</pre>\n",
       "\n",
       "<!-- end doc -->\n",
       "\n",
       "</body></html>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "IPerl->helper( perldoc_html => sub {\n",
    "    shift;\n",
    "    my $html = `perldoc -ohtml Try::Tiny 2> /dev/null`;\n",
    "    return IPerl->html( $html );\n",
    "});\n",
    "\n",
    "IPerl->perldoc_html('Try::Tiny');"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language": "Perl",
  "language_info": {
   "name": "perl",
   "version": "5...1.6...3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
